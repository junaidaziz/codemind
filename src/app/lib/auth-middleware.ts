import { NextRequest, NextResponse } from 'next/server';
import { createServerClient } from './supabase';
import { UserRole, type SessionUser } from '../../types';
import prisma from './db';

// Auth result types
export interface AuthResult {
  success: true;
  user: SessionUser;
}

export interface AuthError {
  success: false;
  error: string;
  code: 'UNAUTHORIZED' | 'FORBIDDEN' | 'INVALID_TOKEN';
  status: number;
}

export type AuthenticationResult = AuthResult | AuthError;

// Authorization options
export interface AuthOptions {
  requiredRole?: UserRole;
  requireAuth?: boolean;
}

/**
 * Middleware to authenticate and authorize requests
 * @param request - Next.js request object
 * @param options - Authentication and authorization options
 * @returns Authentication result with user data or error
 */
export async function authenticateRequest(
  request: NextRequest,
  options: AuthOptions = { requireAuth: true }
): Promise<AuthenticationResult> {
  try {
    const supabase = createServerClient();
    
    // Get session from Supabase
    const { data: { session }, error } = await supabase.auth.getSession();
    
    if (error || !session) {
      return {
        success: false,
        error: 'Authentication required',
        code: 'UNAUTHORIZED',
        status: 401,
      };
    }

    // Get user from our database to include role
    const dbUser = await prisma.user.findUnique({
      where: { id: session.user.id },
      select: {
        id: true,
        email: true,
        name: true,
        image: true,
        role: true,
      },
    });

    if (!dbUser) {
      return {
        success: false,
        error: 'User not found in database',
        code: 'UNAUTHORIZED',
        status: 401,
      };
    }

    // Create typed session user - simplify to avoid complex type conflicts
    const sessionUser = {
      ...session.user,
      email: session.user.email || '',
      role: dbUser.role as UserRole,
    } as SessionUser;

    // Check role-based authorization
    if (options.requiredRole && dbUser.role !== options.requiredRole) {
      return {
        success: false,
        error: `Access denied. Required role: ${options.requiredRole}`,
        code: 'FORBIDDEN',
        status: 403,
      };
    }

    return {
      success: true,
      user: sessionUser,
    };
  } catch (error) {
    console.error('Authentication middleware error:', error);
    return {
      success: false,
      error: 'Authentication failed',
      code: 'INVALID_TOKEN',
      status: 500,
    };
  }
}

/**
 * Higher-order function to protect API routes with authentication
 * @param handler - The API route handler function
 * @param options - Authentication options
 * @returns Protected API route handler
 */
export function withAuth<T extends NextRequest>(
  handler: (req: T, user: SessionUser) => Promise<NextResponse>,
  options: AuthOptions = {}
) {
  return async (req: T): Promise<NextResponse> => {
    const authResult = await authenticateRequest(req, options);
    
    if (!authResult.success) {
      const error = authResult as AuthError;
      return NextResponse.json(
        {
          success: false,
          error: error.error,
          code: error.code,
        },
        { status: error.status }
      );
    }

    return handler(req, authResult.user);
  };
}

/**
 * Middleware to check if user owns a project
 * @param projectId - Project ID to check ownership
 * @param userId - User ID to verify ownership
 * @returns True if user owns the project or is admin
 */
export async function checkProjectOwnership(
  projectId: string,
  userId: string,
  userRole: UserRole
): Promise<boolean> {
  // Admins can access any project
  if (userRole === 'admin') {
    return true;
  }

  try {
    const project = await prisma.project.findUnique({
      where: { id: projectId },
      select: { ownerId: true },
    });

    return project?.ownerId === userId;
  } catch (error) {
    console.error('Error checking project ownership:', error);
    return false;
  }
}

/**
 * Middleware to check if user can access a chat session
 * @param sessionId - Chat session ID to check access
 * @param userId - User ID to verify access
 * @returns True if user can access the session
 */
export async function checkSessionAccess(
  sessionId: string,
  userId: string,
  userRole: UserRole
): Promise<boolean> {
  // Admins can access any session
  if (userRole === 'admin') {
    return true;
  }

  try {
    const session = await prisma.chatSession.findUnique({
      where: { id: sessionId },
      select: { userId: true },
    });

    return session?.userId === userId;
  } catch (error) {
    console.error('Error checking session access:', error);
    return false;
  }
}