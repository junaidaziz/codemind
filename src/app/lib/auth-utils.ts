import { NextRequest } from 'next/server';
import { createServerClient } from './supabase';

export async function getAuthenticatedUser(req: NextRequest) {
  const authHeader = req.headers.get('authorization');
  
  if (!authHeader || !authHeader.startsWith('Bearer ')) {
    return null;
  }

  const token = authHeader.substring(7);
  
  try {
    const supabase = createServerClient();

    const { data: { user }, error } = await supabase.auth.getUser(token);
    
    if (error || !user) {
      return null;
    }

    return user;
  } catch (error) {
    console.error('Auth error:', error);
    return null;
  }
}