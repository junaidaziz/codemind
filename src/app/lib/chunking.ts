import { logger } from './logger';

// Enhanced type definitions for code chunking
export interface CodeChunk {
  path: string;
  content: string;
  startLine: number;
  endLine: number;
  language?: string;
  tokenCount?: number;
}

export interface ChunkingOptions {
  maxLines?: number;
  maxTokens?: number;
  overlapLines?: number;
  preserveFunctions?: boolean;
  minChunkSize?: number;
}

export interface ChunkingResult {
  chunks: CodeChunk[];
  totalLines: number;
  totalChunks: number;
  avgChunkSize: number;
}

/**
 * Enhanced code file chunking with intelligent splitting and overlap
 */
export function chunkCodeFile(
  content: string,
  path: string,
  options: ChunkingOptions = {}
): ChunkingResult {
  const {
    maxLines = 200,
    maxTokens = 1000,
    overlapLines = 20,
    preserveFunctions = true,
    minChunkSize = 10
  } = options;

  const lines = content.split("\n");
  const chunks: CodeChunk[] = [];
  const detectedLanguage = detectLanguage(path);
  
  logger.debug('Starting file chunking', {
    path,
    language: detectedLanguage,
    totalLines: lines.length,
    maxLines,
    maxTokens,
  });

  let start = 0;
  
  while (start < lines.length) {
    const end = findOptimalChunkEnd(
      lines, 
      start, 
      maxLines, 
      detectedLanguage, 
      preserveFunctions
    );
    
    // Skip chunks that are too small
    if (end - start < minChunkSize) {
      start = end;
      continue;
    }
    
    const chunkLines = lines.slice(start, end);
    const chunkContent = chunkLines.join("\n");
    const tokenCount = estimateTokenCount(chunkContent);
    
    // If chunk exceeds token limit, split it further
    if (tokenCount > maxTokens) {
      const subChunks = splitByTokenLimit(chunkLines, start, maxTokens, path, detectedLanguage);
      chunks.push(...subChunks);
    } else {
      chunks.push({
        path,
        content: chunkContent,
        startLine: start + 1,
        endLine: end,
        language: detectedLanguage,
        tokenCount,
      });
    }
    
    // Move start position with overlap
    start = Math.max(end - overlapLines, start + 1);
  }

  const result: ChunkingResult = {
    chunks,
    totalLines: lines.length,
    totalChunks: chunks.length,
    avgChunkSize: chunks.length > 0 ? 
      chunks.reduce((sum, chunk) => sum + (chunk.endLine - chunk.startLine), 0) / chunks.length : 
      0,
  };

  logger.debug('File chunking completed', {
    path,
    language: detectedLanguage,
    ...result,
  });

  return result;
}

/**
 * Detect programming language from file extension
 */
function detectLanguage(path: string): string {
  const ext = path.split('.').pop()?.toLowerCase() || '';
  
  const languageMap: Record<string, string> = {
    'js': 'javascript',
    'jsx': 'javascript',
    'ts': 'typescript',
    'tsx': 'typescript',
    'py': 'python',
    'java': 'java',
    'cpp': 'cpp',
    'c': 'c',
    'cs': 'csharp',
    'php': 'php',
    'rb': 'ruby',
    'go': 'go',
    'rs': 'rust',
    'kt': 'kotlin',
    'swift': 'swift',
    'scala': 'scala',
    'sh': 'bash',
    'sql': 'sql',
    'json': 'json',
    'yaml': 'yaml',
    'yml': 'yaml',
    'xml': 'xml',
    'html': 'html',
    'css': 'css',
    'scss': 'scss',
    'md': 'markdown',
  };
  
  return languageMap[ext] || 'text';
}

/**
 * Find optimal chunk end position considering language constructs
 */
function findOptimalChunkEnd(
  lines: string[],
  start: number,
  maxLines: number,
  language: string,
  preserveFunctions: boolean
): number {
  const maxEnd = Math.min(start + maxLines, lines.length);
  
  if (!preserveFunctions) {
    return maxEnd;
  }
  
  // Language-specific function/class boundary detection
  const patterns = getFunctionPatterns(language);
  
  // Look for natural breaking points (function/class boundaries)
  for (let i = maxEnd - 1; i > start + Math.floor(maxLines * 0.7); i--) {
    const line = lines[i]?.trim() || '';
    
    // Check if this line ends a function/class
    if (patterns.some(pattern => pattern.test(line))) {
      return i + 1;
    }
    
    // Check for blank lines as natural boundaries
    if (line === '' && i > start + Math.floor(maxLines * 0.5)) {
      return i;
    }
  }
  
  return maxEnd;
}

/**
 * Get function/class patterns for different languages
 */
function getFunctionPatterns(language: string): RegExp[] {
  const patterns: Record<string, RegExp[]> = {
    javascript: [/^\s*}\s*$/, /^\s*};\s*$/, /^\s*},?\s*$/, /^\s*}\)\s*$/],
    typescript: [/^\s*}\s*$/, /^\s*};\s*$/, /^\s*},?\s*$/, /^\s*}\)\s*$/],
    python: [/^\s*def\s+\w+/, /^\s*class\s+\w+/, /^\s*async\s+def\s+\w+/],
    java: [/^\s*}\s*$/, /^\s*public\s+class/, /^\s*private\s+class/, /^\s*protected\s+class/],
    cpp: [/^\s*}\s*$/, /^\s*};\s*$/, /^\s*namespace/, /^\s*class/],
    c: [/^\s*}\s*$/, /^\s*};\s*$/, /^\s*struct/, /^\s*typedef/],
  };
  
  return patterns[language] || [/^\s*}\s*$/];
}

/**
 * Split chunk by token limit when it's too large
 */
function splitByTokenLimit(
  lines: string[],
  startLine: number,
  maxTokens: number,
  path: string,
  language: string
): CodeChunk[] {
  const chunks: CodeChunk[] = [];
  let currentStart = 0;
  let currentTokens = 0;
  
  for (let i = 0; i < lines.length; i++) {
    const lineTokens = estimateTokenCount(lines[i]);
    
    if (currentTokens + lineTokens > maxTokens && i > currentStart) {
      // Create chunk from currentStart to i
      const chunkLines = lines.slice(currentStart, i);
      chunks.push({
        path,
        content: chunkLines.join("\n"),
        startLine: startLine + currentStart + 1,
        endLine: startLine + i,
        language,
        tokenCount: currentTokens,
      });
      
      currentStart = i;
      currentTokens = lineTokens;
    } else {
      currentTokens += lineTokens;
    }
  }
  
  // Add remaining lines as final chunk
  if (currentStart < lines.length) {
    const chunkLines = lines.slice(currentStart);
    chunks.push({
      path,
      content: chunkLines.join("\n"),
      startLine: startLine + currentStart + 1,
      endLine: startLine + lines.length,
      language,
      tokenCount: currentTokens,
    });
  }
  
  return chunks;
}

/**
 * Estimate token count for a text (rough approximation)
 */
function estimateTokenCount(text: string): number {
  // Simple heuristic: ~4.5 characters per token for code
  return Math.ceil(text.length / 4.5);
}

/**
 * Batch process multiple files with chunking
 */
export async function chunkMultipleFiles(
  files: { path: string; content: string }[],
  options: ChunkingOptions = {}
): Promise<{
  results: Record<string, ChunkingResult>;
  totalChunks: number;
  processingTime: number;
}> {
  const startTime = Date.now();
  const results: Record<string, ChunkingResult> = {};
  let totalChunks = 0;
  
  logger.info('Starting batch file chunking', {
    fileCount: files.length,
    options,
  });
  
  for (const file of files) {
    const result = chunkCodeFile(file.content, file.path, options);
    results[file.path] = result;
    totalChunks += result.totalChunks;
  }
  
  const processingTime = Date.now() - startTime;
  
  logger.info('Batch file chunking completed', {
    fileCount: files.length,
    totalChunks,
    processingTime,
  });
  
  return {
    results,
    totalChunks,
    processingTime,
  };
}
