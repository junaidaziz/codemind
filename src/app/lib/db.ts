import { PrismaClient } from '@prisma/client';
import { env } from '../../types/env';

const globalForPrisma = globalThis as unknown as {
  prisma: PrismaClient | undefined;
};

const prisma = globalForPrisma.prisma ?? new PrismaClient({
  datasources: {
    db: {
      url: env.DATABASE_URL,
    },
  },
  log: env.NODE_ENV === 'development' ? ['query', 'error', 'warn'] : ['error'],
});

if (env.NODE_ENV !== 'production') globalForPrisma.prisma = prisma;

export default prisma;
