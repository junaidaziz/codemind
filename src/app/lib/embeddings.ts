import OpenAI from "openai";
import { env } from '../../types/env';

const openai = new OpenAI({
  apiKey: env.OPENAI_API_KEY
});

export async function embedTexts(texts: string[]): Promise<number[][]> {
  const res = await openai.embeddings.create({
    model: "text-embedding-3-small",
    input: texts
  });
  return res.data.map(r => r.embedding);
}
