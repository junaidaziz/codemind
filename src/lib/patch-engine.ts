/**
 * Minimal Patch Engine (Phase 3 Stub)
 * Generates and applies extremely simple unified diffs.
 * Future: expand to multi-hunk, context validation, conflict detection.
 */

export interface SimpleDiffPlan {
  filePath: string;
  originalContent: string;
  updatedContent: string;
  unifiedDiff: string; // minimal diff for display
}

export function buildHeaderCommentPatch(filePath: string, source: string): SimpleDiffPlan | null {
  // Avoid double insertion
  if (source.startsWith('// AI: AutoFix header')) return null;
  const header = `// AI: AutoFix header inserted for initial patch planning\n`;
  const updated = header + source;
  const context = parseInt(process.env.AUTOFIX_DIFF_CONTEXT || '', 10);
  const diff = createUnifiedDiffForTest(filePath, source, updated, isNaN(context) ? 3 : Math.max(0, context));
  return { filePath, originalContent: source, updatedContent: updated, unifiedDiff: diff };
}

export function createUnifiedDiffForTest(filePath: string, oldStr: string, newStr: string, context = 3): string {
  const a = oldStr.split('\n');
  const b = newStr.split('\n');
  if (oldStr === newStr) {
    return [
      `--- a/${filePath}`,
      `+++ b/${filePath}`,
      '@@ -0,0 +0,0 @@',
      ' (no changes)'
    ].join('\n');
  }

  // Myers O(ND) diff to produce minimal edit script
  type Op = { type: 'equal' | 'add' | 'del'; line: string; aIndex: number; bIndex: number };
  function myers(aLines: string[], bLines: string[]): Op[] {
    const N = aLines.length; const M = bLines.length;
    const max = N + M;
    const trace: Map<number, number>[] = [];
    let v = new Map<number, number>();
    v.set(1, 0);
    let foundD = 0;
    outer: for (let d = 0; d <= max; d++) {
      const vNext = new Map<number, number>();
      for (let k = -d; k <= d; k += 2) {
        let x: number;
        if (k === -d || (k !== d && (v.get(k - 1) ?? -1) < (v.get(k + 1) ?? -1))) {
          x = v.get(k + 1) ?? 0; // insertion
        } else {
          x = (v.get(k - 1) ?? 0) + 1; // deletion
        }
        let y = x - k;
        while (x < N && y < M && aLines[x] === bLines[y]) { x++; y++; }
        vNext.set(k, x);
        if (x >= N && y >= M) {
          trace.push(vNext);
          foundD = d;
          break outer;
        }
      }
      trace.push(vNext);
      v = vNext;
    }
    // Reconstruct path
    const ops: Op[] = [];
    let x = aLines.length; let y = bLines.length;
    for (let d = foundD; d >= 0; d--) {
      const k = x - y;
      let prevK: number;
      if (d === 0) {
        prevK = 0;
      } else if (k === -d || (k !== d && (trace[d - 1].get(k - 1) ?? -1) < (trace[d - 1].get(k + 1) ?? -1))) {
        prevK = k + 1; // insertion
      } else {
        prevK = k - 1; // deletion
      }
      const vPrev = trace[d - 1];
      const xStart = d === 0 ? 0 : (vPrev.get(prevK) ?? 0);
      const yStart = xStart - prevK;
      while (x > xStart && y > yStart) {
        ops.push({ type: 'equal', line: aLines[x - 1], aIndex: x - 1, bIndex: y - 1 });
        x--; y--;
      }
      if (d === 0) break;
      if (xStart < x) {
        // deletion
        ops.push({ type: 'del', line: aLines[x - 1], aIndex: x - 1, bIndex: y });
        x--;
      } else if (yStart < y) {
        // insertion
        ops.push({ type: 'add', line: bLines[y - 1], aIndex: x, bIndex: y - 1 });
        y--;
      }
    }
    return ops.reverse();
  }

  const ops = myers(a, b);

  // Group into hunks with context
  interface HunkOp { type: 'context' | 'add' | 'del'; line: string }
  interface Hunk { aStart: number; bStart: number; lines: HunkOp[] }
  const hunks: Hunk[] = [];
  let hunk: Hunk | null = null;
  const pushHunk = () => { if (hunk) { hunks.push(hunk); hunk = null; } };
  // Track current indices
  let aIndex = 0; let bIndex = 0;
  for (let idx = 0; idx < ops.length; idx++) {
    const o = ops[idx];
    if (o.type === 'equal') {
      if (hunk) {
        // inside hunk - trailing context management
        if (context > 0) {
          hunk.lines.push({ type: 'context', line: o.line });
          // lookahead for further changes within remaining context window
          let aheadChange = false;
          for (let look = 1; look <= context; look++) {
            const next = ops[idx + look];
            if (!next) break;
            if (next.type !== 'equal') { aheadChange = true; break; }
          }
          if (!aheadChange) pushHunk();
        } else {
          // zero context => close immediately after change block
          pushHunk();
        }
      }
      aIndex++; bIndex++;
    } else {
      // start hunk if none
      if (!hunk) {
        const startA = o.type === 'add' ? aIndex + 1 : aIndex + 1; // 1-based
        const startB = o.type === 'del' ? bIndex + 1 : bIndex + 1;
        hunk = { aStart: startA, bStart: startB, lines: [] };
        // leading context lines
        if (context > 0 && hunks.length === 0) {
          // attempt to include previous context from already processed equals (not stored). Skipped for simplicity.
        }
      }
      if (o.type === 'del') {
        hunk.lines.push({ type: 'del', line: o.line });
        aIndex++;
      } else if (o.type === 'add') {
        hunk.lines.push({ type: 'add', line: o.line });
        bIndex++;
      }
    }
  }
  pushHunk();

  // Add leading context for each hunk retrospectively
  if (context > 0) {
    // We regenerate with another pass using original lines for proper context around hunk boundaries
    for (const h of hunks) {
      const aStart0 = h.aStart - 1; // zero-based
      const preStart = Math.max(0, aStart0 - context);
      const leading: HunkOp[] = [];
      for (let i = preStart; i < aStart0; i++) {
        leading.push({ type: 'context', line: a[i] });
      }
      h.lines = [...leading, ...h.lines];
      // trailing context
      // compute last affected a index
      let lastA = aStart0;
      for (const l of h.lines) {
        if (l.type === 'del' || l.type === 'context') lastA++; // context & del advance in original file perspective
      }
      const trailStart = lastA; // already advanced one past last changed line
      for (let i = trailStart; i < Math.min(a.length, trailStart + context); i++) {
        h.lines.push({ type: 'context', line: a[i] });
      }
    }
  }

  // Merge overlapping context between consecutive hunks
  const merged: Hunk[] = [];
  for (const h of hunks) {
    if (!merged.length) { merged.push(h); continue; }
    const prev = merged[merged.length - 1];
    // crude heuristic: if original line ranges overlap or are adjacent within context*2, merge
    const prevEndA = prev.aStart - 1 + prev.lines.filter(l => l.type !== 'add').length; // approximate span in original
    if (h.aStart <= prevEndA + context * 2) {
      prev.lines.push(...h.lines);
    } else {
      merged.push(h);
    }
  }

  // Build unified diff
  interface Count { added: number; deleted: number; }
  function buildHunk(h: Hunk): { text: string; count: Count } {
    let aLen = 0, bLen = 0, added = 0, deleted = 0;
    for (const l of h.lines) {
      if (l.type === 'context') { aLen++; bLen++; }
      else if (l.type === 'add') { bLen++; added++; }
      else if (l.type === 'del') { aLen++; deleted++; }
    }
    const range = `@@ -${h.aStart},${aLen} +${h.bStart},${bLen} @@`;
    const body = h.lines.map(l => l.type === 'context' ? ` ${l.line}` : (l.type === 'add' ? `+${l.line}` : `-${l.line}`));
    return { text: [range, ...body].join('\n'), count: { added, deleted } };
  }
  const header = [`--- a/${filePath}`, `+++ b/${filePath}`];
  const maxHunks = parseInt(process.env.AUTOFIX_DIFF_MAX_HUNKS || '20', 10) || 20;
  const maxBytes = parseInt(process.env.AUTOFIX_DIFF_MAX_BYTES || '25000', 10) || 25000;
  const built = merged.map(buildHunk);
  const diffBody: string[] = []; let usedBytes = header.join('\n').length; let includedHunks = 0; let totalAdded = 0; let totalDeleted = 0;
  for (const h of built) {
    const projectedSize = usedBytes + h.text.length + 1; // + newline
    if (includedHunks >= maxHunks || projectedSize > maxBytes) break;
    diffBody.push(h.text); usedBytes = projectedSize; includedHunks++; totalAdded += h.count.added; totalDeleted += h.count.deleted;
  }
  const truncated = includedHunks < built.length;
  if (truncated) {
    diffBody.push(`# Diff truncated: showed ${includedHunks}/${built.length} hunks, +${totalAdded} -${totalDeleted} so far.`);
  } else {
    // accumulate totals if not truncated
    for (let i = includedHunks; i < built.length; i++) { totalAdded += built[i].count.added; totalDeleted += built[i].count.deleted; }
  }
  return [...header, ...diffBody].join('\n');
}
