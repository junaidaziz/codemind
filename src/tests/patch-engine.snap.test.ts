import { createUnifiedDiffForTest } from '../lib/patch-engine';

const file = 'snapshot.ts';

describe('Diff snapshots', () => {
  it('single insertion snapshot', () => {
    const oldStr = 'a\nb\nc';
    const newStr = 'a\nX\nb\nc';
    const diff = createUnifiedDiffForTest(file, oldStr, newStr, 1);
    expect(diff).toMatchSnapshot();
  });

  it('multi distant edits snapshot', () => {
    const oldStr = Array.from({length:15}, (_,i)=>`L${i}`).join('\n');
    const lines = oldStr.split('\n');
    lines[2] = lines[2] + '_mod';
    lines[11] = lines[11] + '_mod';
    const diff = createUnifiedDiffForTest(file, oldStr, lines.join('\n'), 1);
    expect(diff).toMatchSnapshot();
  });
});
