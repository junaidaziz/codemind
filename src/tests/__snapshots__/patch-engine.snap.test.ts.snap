// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Diff snapshots multi distant edits snapshot 1`] = `
"--- a/snapshot.ts
+++ b/snapshot.ts
@@ -3,4 +3,4 @@
 L1
-L2
+L2_mod
 L3
 L5
@@ -12,4 +12,4 @@
 L10
-L11
+L11_mod
 L12
 L14"
`;

exports[`Diff snapshots single insertion snapshot 1`] = `
"--- a/snapshot.ts
+++ b/snapshot.ts
@@ -2,2 +2,3 @@
 a
+X
 b"
`;
